lib/srv/app/connections_handler.go

```diff
@@ -0,0 +1,817 @@
+func NewConnectionsHandler(closeContext context.Context, cfg *ConnectionsHandlerConfig) (*ConnectionsHandler, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	awsSigner, err := awsutils.NewSigningService(awsutils.SigningServiceConfig{
+		Clock: cfg.Clock,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	awsHandler, err := appaws.NewAWSSignerHandler(closeContext, appaws.SignerHandlerConfig{
+		SigningService: awsSigner,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	azureHandler, err := appazure.NewAzureHandler(closeContext, appazure.HandlerConfig{
+		Logger: cfg.Logger.With(teleport.ComponentKey, appazure.ComponentKey),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	gcpHandler, err := appgcp.NewGCPHandler(closeContext, appgcp.HandlerConfig{})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	c := &ConnectionsHandler{
+		cfg:          cfg,
+		closeContext: closeContext,
+		awsHandler:   awsHandler,
+		azureHandler: azureHandler,
+		gcpHandler:   gcpHandler,
+		connAuth:     make(map[net.Conn]error),
+		log:          slog.With(teleport.ComponentKey, cfg.ServiceComponent),
+		legacyLogger: logrus.WithFields(logrus.Fields{teleport.ComponentKey: cfg.ServiceComponent}),
+		getAppByPublicAddress: func(ctx context.Context, s string) (types.Application, error) {
+			return nil, trace.NotFound("no applications are being proxied")
+		},
+	}
+
+	// Create a new session cache, this holds sessions that can be used to
+	// forward requests.
+	c.cache, err = utils.NewFnCache(utils.FnCacheConfig{
+		TTL:             5 * time.Minute,
+		Context:         c.closeContext,
+		Clock:           c.cfg.Clock,
+		CleanupInterval: time.Second,
+		OnExpiry:        c.onSessionExpired,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	go c.expireSessions()
+
+	clustername, err := c.cfg.AccessPoint.GetClusterName()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Create and configure HTTP server with authorizing middleware.
+	c.httpServer = c.newHTTPServer(clustername.GetClusterName())
+
+	// TCP server will handle TCP applications.
+	tcpServer, err := c.newTCPServer()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	c.tcpServer = tcpServer
+
+	// Make copy of server's TLS configuration and update it with the specific
+	// functionality this server needs, like requiring client certificates.
+	c.tlsConfig = CopyAndConfigureTLS(c.legacyLogger, c.cfg.AccessPoint, c.cfg.TLSConfig)
+
+	// Figure out the port the proxy is running on.
+	c.proxyPort = c.getProxyPort()
+
+	return c, nil
+}
+
+// SetApplicationsProvider sets the internal state for the monitored applications.
+// This method must be called before the ConnectionsHandler is able to handle connections.
+func (c *ConnectionsHandler) SetApplicationsProvider(fn func(context.Context, string) (types.Application, error)) {
+	c.getAppByPublicAddress = fn
+}
```

Comment:
this is unfortunate. any easy way around it?

Reply:
Yeah, not the most beautiful thing
I tried to not change too many things and extracting the AppServer storage would require much more changes.

I'll merge as is, but I'll try to figure out a better approach in a follow up PR.